File: CSE 2204 - Computer Algorithms Lab/Greedy Algorithms/Fractional_Knapsack.cpp
#include<bits/stdc++.h>
using namespace std;

int main(){
    int capacity, n;
    cout <<"Knapsack Capacity: ";
    cin >> capacity;
    cout <<"Number of Products: ";
    cin >> n;

    int product[n], value[n], weight[n];
    int v_per_w[n];
    for(int i=0; i<n; i++){
        cout <<"Enter Product " << i+1 <<" Value & Weight (in one line): " << endl;
        cin >> value[i] >> weight[i];
        v_per_w[i] = value[i] / weight[i];
    }

    for(int i=0; i<n; i++){
        for(int j=i+1; j<n; j++){
            if(v_per_w[i] < v_per_w[j]) {
                swap(product[i], product[j]);
                swap(value[i], value[j]);
                swap(weight[i], weight[j]);
                swap (v_per_w[i], v_per_w[j]);
            }
        }
    }

    int total_value = 0, current_weight=0;
    for(int i=0; i<n; i++){
        if(current_weight + weight[i]<=capacity){
            current_weight += weight[i];
            total_value += value[i];
        }
        else{
            int fraction = (capacity - current_weight) * v_per_w[i];
            total_value += fraction;
            break;
        }
    }

    cout <<"Maximum Value: " << total_value;

}

File: CSE 2204 - Computer Algorithms Lab/Greedy Algorithms/Fractional_knapsack.py
n=int(input("Enter number of items: "))
c=float(input("Enter knapsack capacity: "))

it=[]
for i in range(1,n+1):
    w,v=map(float,input(f"Enter weight and value of Item {i}: ").split())
    it.append((w,v))

it=sorted([(w,v,v/w) for w,v in it],key=lambda x:x[2],reverse=True)

w=t=0
for a,b,r in it:
    if w+a<=c:
        w+=a; t+=b
    else:
        t+=r*(c-w); break

print("Maximum value =",t)